design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+Bear in mind that once a value has been retrieved, it will automatically be saved in the local `cdk.json` and won't be retrieved again until `cdk context --reset` is called.
+
+Since synthesis-time resolution doesn't create strong coupling between the stacks at the CloudFormation
+level, production and operational issues can arise if the producing stack deletes an exported resource. On the other hand, we hear from customers that strong-referencing behavior of `Fn::ImportValue` is sometimes a curse. Customers tell us that they found themselves stuck with unremovable or stacks that cannot be updates due to imports.
+
+Luckily, we can enable both capabilities. By default, when synthesis-time resolution is used, the CDK will automatically add a resource metadata entry to the template with an `Fn::ImportValue`. This will create the strong coupling between the stacks. Users can opt-out of this behavior by setting `weak: true` when they import the resource.
+
+### Convenience Methods
+
+At the top layer, we will implement a bunch of convenience methods for each AWS resource will provide nice ergonomics for cross-app import/export:
+
+Here's the usage:
+
+```ts
+// this will export "myAwesomeBucket" under the export name "JohnnyBucket"
+myAwesomeBucket.exportBucket('JohnnyBucket');
+
+// now, any CDK app that wishes to refer to this bucket can do this:
+const importedBucket: IBucket = Bucket.importBucket(this, 'JohnnyBucket', {
+  weak: true // weak-reference
+});
+```
+
+> The reason we call this `importBucket` (and `exportBucket`) is because `import` is a reserved word in Java ([#89](https://github.com/awslabs/aws-cdk/issues/89)). Also, in JavaScript both static and instance methods are inherited, so if someone extends `Bucket` (say `ExtraBucket`), we should have a way to distinguish between `ExtraBucket.importBucket` and `ExtraBucket.importExtraBucket`.
+
+The implementation of these two methods is ~trivial:
+
+```ts
+class Bucket {
+  public static importBucket(scope: Construct, exportName: string, options?: ImportOptions): IBucket {
+    return Bucket.deserializeBucket(Stack.find(scope).importObject(exportName, options));
+  }
+
+  public exportBucket(exportName: string): void {
+    Stack.find(this).exportObject(exportName, this);
+  }
+}
+```
+
+### Reference by Physical Name (REQ3)
+
+AWS resources should allow users to reference them by specifying a physical name attribute such as ARN or name:
+
+```ts
+const myBucket = Bucket.fromBucketArn(this, 'arn:aws:s3:::my_bucket');
+const yourBucket = Bucket.fromBucketName(this, 'your_bucket');
+```
+
+Since these methods need to create a new construct, they should utilize the resource's physical name as the construct ID, and also ensure idempotency. Here's an example:
+
+```ts
+public static fromBucketArn(scope: Construct, bucketArn: string): IBucket {
+  const stack = Stack.find(scope);
+  const id = `fromBucketArn:${bucketArn}`;
```

Comment:
That's where I think we shouldn't try so hard to ensure single-reference.

```ts
const bucketA = Bucket.fromBucketArn(this, 'arn:aws:s3:::BucketName');
const bucketB = Bucket.fromBucketArn(this, stack.formatArn({ service: 's3', resource: 'BucketName', account: '', region: '' }));
const bucketC = Bucket.fromBucketArn(this, 'BucketName');

bucketA === bucketB; // => true
bucketA === bucketC; // => false, WTF?
```

To me, if you cannot guarantee all imports of a given entity result in the same instance, then the "best effort" attempt to reduce churn shouldn't be normative and doesn't warrant being documented so much.

Reply:
I've relaxed the requirement to "it can be the same instance". Is that okay?